### PR TITLE
Allow table.copy between tables with different index types

### PR DESCRIPTION
**Pull request by sbc100**
See https://github.com/WebAssembly/memory64/pull/7 See #6

### CODE REVIEW

File: interpreter/syntax/types.ml [hunk 1]
```diff
@@ -106,6 +106,10 @@ let match_extern_type et1 et2 =
   | ExternGlobalType gt1, ExternGlobalType gt2 -> match_global_type gt1 gt2
   | _, _ -> false
 
+let min_index_type it1 it2 =
```
Comment (rossberg) [inline interpreter/syntax/types.ml]: I don't think this function is needed, you ought to be able to just use OCaml's polymorphic `min` function.

File: interpreter/valid/valid.ml [hunk 2]
```diff
@@ -340,13 +340,11 @@ let rec check_instr (c : context) (e : instr) (s : infer_result_type) : op_type
   | TableCopy (x, y) ->
     let TableType (_lim1, it1, t1) = table c x in
     let TableType (_lim2, it2, t2) = table c y in
+    let it3 = min_index_type it1 it2 in
```
Comment (rossberg) [inline interpreter/valid/valid.ml]:
```suggestion
    let it3 = min it1 it2 in
```

**Review by rossberg — APPROVED** [review]
LGTM modulo comment

File: interpreter/syntax/types.ml (hunk 1 above)
Comment (sbc100) [reply interpreter/syntax/types.ml]: Interesting. That does seem to work. 

I though I32IndexType and I64IndexType were just essentially enum values that make up the `index_type` type.  How does ocaml know how to compile them in functions like min?

Comment (rossberg) [reply interpreter/syntax/types.ml]: OCaml uses a uniform representation, so can have a single polymorphic implementation for an operator like `<=` that works for arbitrary values. Though here the operand type is monomorphic and compiled to int, so the compiler probably specialises that to an int comparison after inlining `min`.

Comment (sbc100) [reply interpreter/syntax/types.ml]: But I don't see `I32IndexType` and `I64IndexType` defined as integer constants anywhere.    I must be missing something obvious here..  Is ocaml simply assigned numbers to them like an enum in C?

Comment (rossberg) [reply interpreter/syntax/types.ml]: It compiles them to numbers in order of appearance. But unlike in C that's just an implementation detail and not observable, only their order is.

**Pull request merged**